File: java/src/main/java/bouchonnois/service/PartieDeChasseService.java
package bouchonnois.service;

import bouchonnois.domain.Chasseur;
import bouchonnois.domain.Event;
import bouchonnois.domain.PartieDeChasse;
import bouchonnois.domain.PartieStatus;
import bouchonnois.repository.PartieDeChasseRepository;
import bouchonnois.service.exceptions.*;
import io.vavr.Tuple2;
import lombok.AllArgsConstructor;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.UUID;
import java.util.function.Supplier;
import java.util.stream.Collectors;

@AllArgsConstructor
public class PartieDeChasseService {
    private final PartieDeChasseRepository repository;
    private final Supplier<LocalDateTime> timeProvider;

    public UUID démarrer(Tuple2<String, Integer> terrainDeChasse, List<Tuple2<String, Integer>> chasseurs) throws ImpossibleDeDémarrerUnePartieSansGalinettes, ImpossibleDeDémarrerUnePartieAvecUnChasseurSansBalle, ImpossibleDeDémarrerUnePartieSansChasseur {
        if (terrainDeChasse._2 <= 0) {
            throw new ImpossibleDeDémarrerUnePartieSansGalinettes();
        }

        PartieDeChasse partieDeChasse = new PartieDeChasse();
        partieDeChasse.setId(UUID.randomUUID());
        partieDeChasse.setStatus(PartieStatus.EN_COURS);
        partieDeChasse.setChasseurs(new ArrayList<>());

        Terrain terrain = new Terrain(terrainDeChasse._1);
        terrain.setNbGalinettes(terrainDeChasse._2);
        partieDeChasse.setTerrain(terrain);
        partieDeChasse.setEvents(new ArrayList<>());

        for (var chasseur : chasseurs) {
            if (chasseur._2 == 0) {
                throw new ImpossibleDeDémarrerUnePartieAvecUnChasseurSansBalle();
            }
            Chasseur chasseurToAdd = new Chasseur();
            chasseurToAdd.setNom(chasseur._1);
            chasseurToAdd.setBallesRestantes(chasseur._2);
            partieDeChasse.getChasseurs().add(chasseurToAdd);
        }

        if (partieDeChasse.getChasseurs().isEmpty()) {
            throw new ImpossibleDeDémarrerUnePartieSansChasseur();
        }

        String chasseursToString = String.join(", ",
                partieDeChasse.getChasseurs().stream()
                        .map(c -> c.getNom() + " (" + c.getBallesRestantes() + " balles)")
                        .toArray(String[]::new));

        partieDeChasse.getEvents().add(new Event(timeProvider.get(),
                "La partie de chasse commence à " + partieDeChasse.getTerrain().getNom() + " avec " + chasseursToString));

        repository.save(partieDeChasse);

        return partieDeChasse.getId();
    }

    public void tirerSurUneGalinette(UUID id, String chasseur) throws LaPartieDeChasseNexistePas, TasPlusDeBallesMonVieuxChasseALaMain, ChasseurInconnu, OnTirePasPendantLapéroCestSacré, TasTropPicoledMonVieuxTasRienTouche, OnTirePasQuandLaPartieEstTerminee {
        PartieDeChasse partieDeChasse = repository.getById(id);

        if (partieDeChasse == null) {
            throw new LaPartieDeChasseNexistePas();
        }

        if (partieDeChasse.getTerrain().getNbGalinettes() != 0) {

            if (partieDeChasse.getStatus() != PartieStatus.APÉRO) {
                if (partieDeChasse.getStatus() != PartieStatus.TERMINÉE) {
                    if (partieDeChasse.getChasseurs().stream().anyMatch(c -> c.getNom().equals(chasseur))) {
                        var chasseurQuiTire = partieDeChasse.getChasseurs().stream()
                                .filter(c -> c.getNom().equals(chasseur)).findFirst()
                                .get();

                        if (chasseurQuiTire.getBallesRestantes() == 0) {
                            partieDeChasse.getEvents().add(new Event(timeProvider.get(),
                                    chasseur + " veut tirer sur une galinette -> T'as plus de balles mon vieux, chasse à la main"));
                            repository.save(partieDeChasse);

                            throw new TasPlusDeBallesMonVieuxChasseALaMain();
                        }

                        chasseurQuiTire.setBallesRestantes(chasseurQuiTire.getBallesRestantes() - 1);
                        chasseurQuiTire.setNbGalinettes(chasseurQuiTire.getNbGalinettes() + 1);
                        partieDeChasse.getTerrain().setNbGalinettes(partieDeChasse.getTerrain().getNbGalinettes() - 1);
                        partieDeChasse.getEvents().add(new Event(timeProvider.get(), chasseur + " tire sur une galinette"));

                    } else {
                        throw new ChasseurInconnu(chasseur);
                    }
                } else {
                    partieDeChasse.getEvents().add(new Event(timeProvider.get(), chasseur + " veut tirer -> On tire pas quand la partie est terminée"));
                    repository.save(partieDeChasse);

                    throw new OnTirePasQuandLaPartieEstTerminee();
                }
            } else {
                partieDeChasse.getEvents()
                        .add(new Event(timeProvider.get(), chasseur + " veut tirer -> On tire pas pendant l'apéro, c'est sacré !!!"));
                repository.save(partieDeChasse);
                throw new OnTirePasPendantLapéroCestSacré();
            }
        } else {
            throw new TasTropPicoledMonVieuxTasRienTouche();
        }

        repository.save(partieDeChasse);
    }

    public void tirer(UUID id, String chasseur) throws LaPartieDeChasseNexistePas, TasPlusDeBallesMonVieuxChasseALaMain, ChasseurInconnu, OnTirePasPendantLapéroCestSacré, TasTropPicoledMonVieuxTasRienTouche, OnTirePasQuandLaPartieEstTerminee {
        PartieDeChasse partieDeChasse = repository.getById(id);

        if (partieDeChasse == null) {
            throw new LaPartieDeChasseNexistePas();
        }
        if (partieDeChasse.getStatus() != PartieStatus.APÉRO) {
            if (partieDeChasse.getStatus() != PartieStatus.TERMINÉE) {
                if (partieDeChasse.getChasseurs().stream().anyMatch(c -> c.getNom().equals(chasseur))) {
                    var chasseurQuiTire = partieDeChasse.getChasseurs()
                            .stream()
                            .filter(c -> c.getNom().equals(chasseur))
                            .findFirst()
                            .get();

                    if (chasseurQuiTire.getBallesRestantes() == 0) {
                        partieDeChasse.getEvents().add(new Event(timeProvider.get(),
                                chasseur + " tire -> T'as plus de balles mon vieux, chasse à la main"));
                        repository.save(partieDeChasse);

                        throw new TasPlusDeBallesMonVieuxChasseALaMain();
                    }

                    chasseurQuiTire.setBallesRestantes(chasseurQuiTire.getBallesRestantes() - 1);
                    partieDeChasse.getEvents().add(new Event(timeProvider.get(), chasseur + " tire"));

                } else {
                    throw new ChasseurInconnu(chasseur);
                }
            } else {
                partieDeChasse.getEvents().add(new Event(timeProvider.get(), chasseur + " veut tirer -> On tire pas quand la partie est terminée"));
                repository.save(partieDeChasse);

                throw new OnTirePasQuandLaPartieEstTerminee();
            }
        } else {
            partieDeChasse.getEvents().add(new Event(timeProvider.get(), chasseur + " veut tirer -> On tire pas pendant l'apéro, c'est sacré !!!"));
            repository.save(partieDeChasse);
            throw new OnTirePasPendantLapéroCestSacré();
        }
        repository.save(partieDeChasse);
    }

    public void prendreLapéro(UUID id) throws LaPartieDeChasseNexistePas, OnEstDéjaEnTrainDePrendreLapéro, OnPrendPasLapéroQuandLaPartieEstTerminée {
        PartieDeChasse partieDeChasse = repository.getById(id);

        if (partieDeChasse == null) {
            throw new LaPartieDeChasseNexistePas();
        }

        if (partieDeChasse.getStatus() == PartieStatus.APÉRO) {
            throw new OnEstDéjaEnTrainDePrendreLapéro();
        } else if (partieDeChasse.getStatus() == PartieStatus.TERMINÉE) {
            throw new OnPrendPasLapéroQuandLaPartieEstTerminée();
        } else {
            partieDeChasse.setStatus(PartieStatus.APÉRO);
            partieDeChasse.getEvents().add(new Event(timeProvider.get(), "Petit apéro"));
            repository.save(partieDeChasse);
        }
    }

    public void reprendreLaPartie(UUID id) throws LaPartieDeChasseNexistePas, LaChasseEstDéjaEnCours, QuandCestFiniCestFini {
        PartieDeChasse partieDeChasse = repository.getById(id);

        if (partieDeChasse == null) {
            throw new LaPartieDeChasseNexistePas();
        }

        if (partieDeChasse.getStatus() == PartieStatus.EN_COURS) {
            throw new LaChasseEstDéjaEnCours();
        }

        if (partieDeChasse.getStatus() == PartieStatus.TERMINÉE) {
            throw new QuandCestFiniCestFini();
        }

        partieDeChasse.setStatus(PartieStatus.EN_COURS);
        partieDeChasse.getEvents().add(new Event(timeProvider.get(), "Reprise de la chasse"));
        repository.save(partieDeChasse);
    }

    public String terminerLaPartie(UUID id) throws QuandCestFiniCestFini {
        PartieDeChasse partieDeChasse = repository.getById(id);

        var classement = partieDeChasse
                .getChasseurs()
                .stream()
                .collect(Collectors.groupingBy(Chasseur::getNbGalinettes));

        if (partieDeChasse.getStatus() == PartieStatus.TERMINÉE) {
            throw new QuandCestFiniCestFini();
        }

        partieDeChasse.setStatus(PartieStatus.TERMINÉE);

        String result = "";

        if (classement.entrySet().stream().allMatch(entry -> entry.getKey() == 0)) {
            result = "Brocouille";
            partieDeChasse.getEvents().add(new Event(timeProvider.get(), "La partie de chasse est terminée, vainqueur : Brocouille"));
        } else {
            var winners = classement.get(classement.keySet().stream()
                    .max(Integer::compare)
                    .orElse(0));

            result = winners.stream()
                    .map(Chasseur::getNom)
                    .collect(Collectors.joining(", "));

            partieDeChasse.getEvents().add(new Event(timeProvider.get(), "La partie de chasse est terminée, vainqueur : " +
                    winners.stream()
                            .map(chasseur -> chasseur.getNom() + " - " + chasseur.getNbGalinettes() + " galinettes")
                            .collect(Collectors.joining(", "))));
        }
        repository.save(partieDeChasse);

        return result;
    }

    public String consulterStatus(UUID id) throws LaPartieDeChasseNexistePas {
        PartieDeChasse partieDeChasse = repository.getById(id);

        if (partieDeChasse == null) {
            throw new LaPartieDeChasseNexistePas();
        }

        List<Event> events = partieDeChasse.getEvents();
        events.sort(Comparator.comparing(Event::date).reversed());

        return events.stream()
                .map(Event::toString)
                .collect(Collectors.joining(System.lineSeparator()));
    }
}

File: kotlin/refactoring-du-bouchonnois/src/main/kotlin/bouchonnois/service/PartieDeChasseService.kt
package bouchonnois.service

import bouchonnois.domain.Chasseur
import bouchonnois.domain.Event
import bouchonnois.domain.PartieDeChasse
import bouchonnois.domain.PartieStatus
import bouchonnois.repository.PartieDeChasseRepository
import bouchonnois.service.exceptions.*
import java.time.LocalDateTime
import java.util.*

class PartieDeChasseService(
    private val repository: PartieDeChasseRepository,
    private val timeProvider: () -> LocalDateTime
) {
    fun démarrer(terrainDeChasse: Pair<String, Int>, chasseurs: Map<String, Int>): UUID {
        if (terrainDeChasse.second <= 0) {
            throw ImpossibleDeDémarrerUnePartieSansGalinettes()
        }

        val partieDeChasse = PartieDeChasse()
        partieDeChasse.id = UUID.randomUUID()
        partieDeChasse.status = PartieStatus.EN_COURS
        partieDeChasse.chasseurs = mutableListOf()

        val terrain = Terrain(terrainDeChasse.first)
        terrain.nbGalinettes = terrainDeChasse.second

        partieDeChasse.terrain = terrain
        partieDeChasse.events = mutableListOf()

        for (chasseur in chasseurs) {
            if (chasseur.value == 0) {
                throw ImpossibleDeDémarrerUnePartieAvecUnChasseurSansBalle()
            }
            val chasseurToAdd = Chasseur()
            chasseurToAdd.nom = chasseur.key
            chasseurToAdd.ballesRestantes = chasseur.value

            partieDeChasse.chasseurs!!.add(chasseurToAdd)
        }

        if (partieDeChasse.chasseurs!!.isEmpty()) {
            throw ImpossibleDeDémarrerUnePartieSansChasseur()
        }

        val chasseursToString = partieDeChasse
            .chasseurs!!
            .joinToString(", ") { "${it.nom} (${it.ballesRestantes} balles)" }

        partieDeChasse.events!!.add(
            Event(
                timeProvider(),
                "La partie de chasse commence à ${partieDeChasse.terrain!!.nom} avec $chasseursToString"
            )
        )
        repository.save(partieDeChasse)

        return partieDeChasse.id!!
    }

    fun tirerSurUneGalinette(id: UUID, chasseur: String) {
        val partieDeChasse = repository.getById(id) ?: throw LaPartieDeChasseNexistePas()

        if (partieDeChasse.terrain!!.nbGalinettes != 0) {
            if (partieDeChasse.status != PartieStatus.APÉRO) {
                if (partieDeChasse.status != PartieStatus.TERMINÉE) {
                    if (partieDeChasse.chasseurs!!.any { c -> c.nom == chasseur }) {
                        val chasseurQuiTire = partieDeChasse.chasseurs!!.first { c -> c.nom == chasseur }

                        if (chasseurQuiTire.ballesRestantes == 0) {
                            partieDeChasse.events!!.add(
                                Event(
                                    timeProvider(),
                                    "$chasseur veut tirer sur une galinette -> T'as plus de balles mon vieux, chasse à la main"
                                )
                            )
                            repository.save(partieDeChasse)
                            throw TasPlusDeBallesMonVieuxChasseALaMain()
                        }
                        chasseurQuiTire.ballesRestantes -= 1
                        chasseurQuiTire.nbGalinettes += 1
                        partieDeChasse.terrain!!.nbGalinettes -= 1
                        partieDeChasse.events!!.add(
                            Event(
                                timeProvider(),
                                "$chasseur tire sur une galinette"
                            )
                        )
                    } else {
                        throw ChasseurInconnu(chasseur)
                    }
                } else {
                    partieDeChasse.events!!.add(
                        Event(
                            timeProvider(),
                            "$chasseur veut tirer -> On tire pas quand la partie est terminée"
                        )
                    )
                    repository.save(partieDeChasse)
                    throw OnTirePasQuandLaPartieEstTerminee()
                }
            } else {
                partieDeChasse
                    .events!!.add(
                        Event(
                            timeProvider(),
                            "$chasseur veut tirer -> On tire pas pendant l'apéro, c'est sacré !!!"
                        )
                    )
                repository.save(partieDeChasse)
                throw OnTirePasPendantLapéroCestSacré()
            }
        } else {
            throw TasTropPicoledMonVieuxTasRienTouche()
        }

        repository.save(partieDeChasse)
    }

    fun tirer(id: UUID, chasseur: String) {
        val partieDeChasse = repository.getById(id) ?: throw LaPartieDeChasseNexistePas()

        if (partieDeChasse.status != PartieStatus.APÉRO) {
            if (partieDeChasse.status != PartieStatus.TERMINÉE) {
                if (partieDeChasse.chasseurs!!.any { c -> c.nom == chasseur }) {
                    val chasseurQuiTire = partieDeChasse.chasseurs!!.first { c -> c.nom == chasseur }

                    if (chasseurQuiTire.ballesRestantes == 0) {
                        partieDeChasse.events!!.add(
                            Event(
                                timeProvider(),
                                "$chasseur tire -> T'as plus de balles mon vieux, chasse à la main"
                            )
                        )
                        repository.save(partieDeChasse)
                        throw TasPlusDeBallesMonVieuxChasseALaMain()
                    }
                    chasseurQuiTire.ballesRestantes = chasseurQuiTire.ballesRestantes - 1
                    partieDeChasse.events!!.add(Event(timeProvider(), "$chasseur tire"))
                } else {
                    throw ChasseurInconnu(chasseur)
                }
            } else {
                partieDeChasse.events!!.add(
                    Event(
                        timeProvider(),
                        "$chasseur veut tirer -> On tire pas quand la partie est terminée"
                    )
                )
                repository.save(partieDeChasse)
                throw OnTirePasQuandLaPartieEstTerminee()
            }
        } else {
            partieDeChasse.events!!.add(
                Event(
                    timeProvider(),
                    "$chasseur veut tirer -> On tire pas pendant l'apéro, c'est sacré !!!"
                )
            )
            repository.save(partieDeChasse)
            throw OnTirePasPendantLapéroCestSacré()
        }
        repository.save(partieDeChasse)
    }

    fun prendreLapéro(id: UUID) {
        val partieDeChasse = repository.getById(id) ?: throw LaPartieDeChasseNexistePas()

        if (partieDeChasse.status == PartieStatus.APÉRO) {
            throw OnEstDéjaEnTrainDePrendreLapéro()
        } else if (partieDeChasse.status == PartieStatus.TERMINÉE) {
            throw OnPrendPasLapéroQuandLaPartieEstTerminée()
        } else {
            partieDeChasse.status = PartieStatus.APÉRO
            partieDeChasse.events!!.add(Event(timeProvider(), "Petit apéro"))
            repository.save(partieDeChasse)
        }
    }

    fun reprendreLaPartie(id: UUID) {
        val partieDeChasse = repository.getById(id)

        if (partieDeChasse == null) {
            throw LaPartieDeChasseNexistePas()
        }
        if (partieDeChasse.status == PartieStatus.EN_COURS) {
            throw LaChasseEstDéjaEnCours()
        }
        if (partieDeChasse.status == PartieStatus.TERMINÉE) {
            throw QuandCestFiniCestFini()
        }

        partieDeChasse.status = PartieStatus.EN_COURS
        partieDeChasse.events!!.add(Event(timeProvider(), "Reprise de la chasse"))
        repository.save(partieDeChasse)
    }

    fun terminerLaPartie(id: UUID): String {
        val partieDeChasse = repository.getById(id)

        val classement = partieDeChasse
            ?.chasseurs
            ?.groupBy { c -> c.nbGalinettes }

        if (partieDeChasse!!.status == PartieStatus.TERMINÉE) {
            throw QuandCestFiniCestFini()
        }

        partieDeChasse.status = PartieStatus.TERMINÉE

        val result: String

        if (classement!!.keys.all { it == 0 }) {
            result = "Brocouille"
            partieDeChasse.events!!.add(
                Event(
                    timeProvider(),
                    "La partie de chasse est terminée, vainqueur : Brocouille"
                )
            )
        } else {
            val winners = classement[classement.keys.max()]
            result = winners!!
                .map(Chasseur::nom)
                .joinToString(", ")

            partieDeChasse.events!!.add(
                Event(
                    timeProvider(),
                    "La partie de chasse est terminée, vainqueur : ${
                        winners.map { chasseur: Chasseur -> chasseur.nom + " - " + chasseur.nbGalinettes + " galinettes" }
                    }"
                ))
        }
        repository.save(partieDeChasse)

        return result
    }

    fun consulterStatus(id: UUID): String {
        val partieDeChasse = repository.getById(id)
        if (partieDeChasse == null) throw LaPartieDeChasseNexistePas()

        return partieDeChasse.events!!
            .sortedByDescending { it.date }
            .joinToString(System.lineSeparator()) { it.toString() }
    }
}